network/src/lib.rs

```diff
@@ -148,6 +148,38 @@ where
 	}
 }
 
+pub struct DelayedBlockAnnounceValidator<B: BlockT>(Arc<Mutex<Option<Box<dyn BlockAnnounceValidator<B> + Send>>>>);
+
+impl<B: BlockT> DelayedBlockAnnounceValidator<B> {
+	pub fn new() -> DelayedBlockAnnounceValidator<B> {
+		DelayedBlockAnnounceValidator(Arc::new(Mutex::new(None)))
+	}
+
+	pub fn set(&self, validator: Box<dyn BlockAnnounceValidator<B> + Send>) {
+		*self.0.lock().unwrap() = Some(validator);
+	}
+}
+
+impl<B: BlockT> Clone for DelayedBlockAnnounceValidator<B> {
+	fn clone(&self) -> DelayedBlockAnnounceValidator<B> {
+		DelayedBlockAnnounceValidator(self.0.clone())
+	}
+}
+
+impl<B: BlockT> BlockAnnounceValidator<B> for DelayedBlockAnnounceValidator<B>
+{
+	fn validate(
+		&mut self,
+		header: &B::Header,
+		data: &[u8],
+	) -> Result<Validation, Box<dyn std::error::Error + Send>> {
+		match *self.0.lock().unwrap() {
+			Some(ref mut validator) => validator.validate(header, data),
```

Comment:
You can just use `as_mut().expect("`BlockAnnounceValidator` is set before validating the first announcement; qed")` to remove this match.